src/Adapter/Product/CommandHandler/GenerateProductCombinationsHandler.php

```diff
@@ -0,0 +1,194 @@
+{
+    /**
+     * @var CombinationGeneratorInterface
+     */
+    private $combinationGenerator;
+
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var Db
+     */
+    private $dbInstance;
+
+    /**
+     * @param CombinationGeneratorInterface $combinationGenerator
+     * @param Connection $connection
+     */
+    public function __construct(
+        CombinationGeneratorInterface $combinationGenerator,
+        Connection $connection
+    ) {
+        $this->combinationGenerator = $combinationGenerator;
+        $this->connection = $connection;
+        $this->dbInstance = Db::getInstance();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @todo: multistore
+     */
+    public function handle(GenerateProductCombinationsCommand $command): array
+    {
+        $product = $this->getProduct($command->getProductId());
+        $generatedCombinations = $this->combinationGenerator->generate($command->getGroupedAttributeIds());
+
+        SpecificPriceRule::disableAnyApplication();
+
+        $combinationIds = $this->addCombinations($product, $generatedCombinations);
+
+        Product::updateDefaultAttribute($product->id);
+        SpecificPriceRule::enableAnyApplication();
+        SpecificPriceRule::applyAllRules([$product->id]);
+
+        return $combinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param Iterator $generatedCombinations
+     *
+     * @return CombinationId[]
+     *
+     * @throws CombinationException
+     */
+    private function addCombinations(Product $product, Iterator $generatedCombinations): array
+    {
+        $addedCombinationIds = [];
+        foreach ($generatedCombinations as $generatedCombination) {
+            $combinationExists = $product->productAttributeExists($generatedCombination, false, null, true);
+
+            if ($combinationExists) {
+                continue;
+            }
+
+            try {
+                $addedCombinationIds[] = $this->addCombination($product, $generatedCombination);
+            } catch (PrestaShopException $e) {
+                throw new CombinationException(
+                    sprintf(
+                        'Error occurred when trying to add combination for product #%d. Combination: %s',
+                        $product->id,
+                        var_export($generatedCombination, true)
+                    ),
+                    $e
+                );
+            }
+        }
+
+        return $addedCombinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param int[] $generatedCombination
+     *
+     * @return CombinationId
+     */
+    private function addCombination(Product $product, array $generatedCombination): CombinationId
+    {
+        $newCombination = new Combination();
+        $newCombination->id_product = $product->id;
+        $newCombination->default_on = 0;
+
+        $this->connection->beginTransaction();
```

Comment:
Using a transaction is a great idea, however two things are bothering me:
- to have a better impact on performance shouldn't the transaction begin at the start of the loop and commit at the end, here it is only used for One combination addition so it's not such a huge query
- I'm surprised that Doctrine connection and the legacy Db class or ObjectModel add/save methods are related? To really profit from the transaction I think the insertion request need to be performed via this Doctrine connection no?

Reply:
I think transaction checks sql queries that was done, it doesnt care about php part. :thinking:

Reply:
I've tested it, indeed it didn't work. Changed to use Dbinstance transaction, works as expected now.